modules/affinity/src/main/java/org/apache/ignite/internal/affinity/Assignments.java

```diff
@@ -0,0 +1,228 @@
+     *
+     * @param bytes Byte array representation of the assignments.
+     * @return Assignments.
+     */
+    public static Assignments fromBytes(byte[] bytes) {
+        ByteBuffer buf = ByteBuffer.wrap(bytes);
+
+        int consistentIdsBytesLength = buf.getInt();
+        byte[] consistentIdsBytes = new byte[consistentIdsBytesLength];
+        buf.get(consistentIdsBytes);
+
+        int assignmentsBytesLength = buf.getInt();
+        byte[] assignmentsBytes = new byte[assignmentsBytesLength];
+        buf.get(assignmentsBytes);
+
+        List<String> consistentIds = bytesToList(consistentIdsBytes, b -> new String(b, StandardCharsets.UTF_8));
+
+        List<Set<Assignment>> assignments = bytesToList(
+                assignmentsBytes,
+                b -> new HashSet<>(bytesToList(b, ab -> bytesToAssignment(ab, consistentIds::get)))
+        );
+
+        return new Assignments(assignments);
+    }
+
+    /**
+     * Serializes collection to bytes.
+     *
+     * @param collection Collection.
+     * @param transform Tranform function for the collection element.
+     * @return Byte array.
+     */
+    private <T> byte[] collectionToBytes(Collection<T> collection, Function<T, byte[]> transform) {
+        int bytesObjects = 0;
+        List<byte[]> objects = new ArrayList<>();
+
+        for (T o : collection) {
+            byte[] b = transform.apply(o);
+            objects.add(b);
+            bytesObjects += b.length;
+        }
+
+        bytesObjects += Short.BYTES * (objects.size() + 1);
+
+        ByteBuffer buf = ByteBuffer.allocate(bytesObjects);
+
+        buf.putShort((short) objects.size());
+
+        for (byte[] o : objects) {
+            buf.putShort((short) o.length);
+            buf.put(o);
+        }
+
+        return buf.array();
+    }
+
+    /**
+     * Deserializes the list from byte array.
+     *
+     * @param bytes Byte array.
+     * @param transform Transform function to create list element.
+     * @return List.
+     */
+    private static <T> List<T> bytesToList(byte[] bytes, Function<byte[], T> transform) {
+        ByteBuffer buf = ByteBuffer.wrap(bytes);
+
+        short length = buf.getShort();
+        assert length >= 0 : "Negative collection size: " + length;
+
+        List<T> result = new ArrayList<>(length);
+
+        for (int i = 0; i < length; i++) {
+            short size = buf.getShort();
+            assert size >= 0 : "Negative object size: " + size + ", index=" + i;
+            byte[] arr = new byte[size];
+            buf.get(arr);
+            result.add(transform.apply(arr));
+        }
+
+        return result;
+    }
+
+    private static byte[] assignmentToBytes(Assignment assignment, Function<String, Integer> consistentIdToIndex) {
+        ByteBuffer buf = ByteBuffer.allocate(Integer.BYTES + Byte.BYTES);
+        buf.putInt(consistentIdToIndex.apply(assignment.consistentId()));
+        buf.put((byte) (assignment.isPeer() ? 1 : 0));
```

Comment:
Seems a bit wasteful. We could encode the "peer" bit in the consistent id itself.
By the way, do we use this class for CMG/Meta-Storage or not? They are replicated, in a nutshell, so this encoding would result in huge payloads anyway